SurfaceTopography/NonuniformLineScan.py

```diff
@@ -132,41 +164,91 @@ def squeeze(self):
+
+        Arguments
+        ---------
+        topography : :obj:`NonuniformTopographyInterface`
+            Parent topography
+        height_scale_factor : float
+            Factor to scale heights with. If None, scale factor will be
+            autodetermined from the given unit. (Default: None)
+        position_scale_factor : float
+            Factor to scale lateral positions (`physical_sizes`, etc.) with.
+            If None, scale factor will be autodetermined from the given unit.
+            (Default: None)
+        unit : str
+            Target unit. Ignored if None. If it is given, then
+            `heights_scale_factor` and `positions_scale_factor` are determined
+            from the unit, if they are None. The `info` dictionary is updated
+            with this unit information. If scale factors are given, they
+            override the autodetection, i.e. if both scale factors are given
+            the unit will just update the `info` dictionary.
+        info : dict
+            Updated entries to the info dictionary.
         """
-        super().__init__(topography, info=info)
-        self._scale_factor = float(scale_factor)
+        super().__init__(topography, unit=unit, info=info)
+        self._height_scale_factor = None if height_scale_factor is None else float(height_scale_factor)
+        self._position_scale_factor = None if position_scale_factor is None else float(position_scale_factor)
 
     def __getstate__(self):
         """ is called and the returned object is pickled as the contents for
             the instance
         """
-        state = super().__getstate__(), self._scale_factor
+        state = super().__getstate__(), self._height_scale_factor, self._position_scale_factor, self._unit
         return state
 
     def __setstate__(self, state):
         """ Upon unpickling, it is called with the unpickled state
         Keyword Arguments:
         state -- result of __getstate__
         """
-        superstate, self._scale_factor = state
+        superstate, self._height_scale_factor, self._position_scale_factor, self._unit = state
         super().__setstate__(superstate)
 
     @property
+    def height_scale_factor(self):
+        if self._height_scale_factor is None:
+            return get_unit_conversion_factor(self.parent_topography.unit, self.unit)
+        else:
+            return self._height_scale_factor
+
+    # For backwards compatibility
+    @property
+    @deprecated(version='0.94.0', reason='Use heights_scale_factor')
     def scale_factor(self):
-        return self._scale_factor
+        return self.height_scale_factor
+
+    @property
+    def position_scale_factor(self):
+        if self._position_scale_factor is None:
+            if self.parent_topography.unit is None:
+                if self.unit is not None:
+                    raise ValueError('You are trying to rescale a unitless topography to a topography with a unit')
```

Comment:
The user just requests the scale factor for positions, why is there an exception saying that the user tries to rescale?

If this is a check whether the configuration of the topography is valid, this should be done in the constructor,
because otherwise the topography can be used although it is invalid.

Reply:
What is meant by this error is the following: You are running `t.scale(unit='um')`, i.e. you want to rescale to 'um' but `t` does not have a unit. This is only possible if `t` has a unit. I made the error clearer.

Reply:
I've also added the error to the constructor

Reply:
(Although the parent topography may change. So constructor may not be able to decide this.)

Reply:
(This is in theory, in the current implementation of the code this cannot happen.)

Reply:
> What is meant by this error is the following: You are running `t.scale(unit='um')`, i.e. you want to rescale to 'um' but `t` does not have a unit. This is only possible if `t` has a unit. I made the error clearer.

But the exception is thrown in the property `position_scale_factor`, not in the `scale` method?
However, it is now also in the constructor and that should be sufficient.

Maybe we could do this kind of check for all height containers automatically.